app/scripts/lib/stores/persistence-manager.ts

```diff
@@ -57,58 +65,138 @@ export class PersistenceManager {
    * includes a single key which is 'version' and contains the current version
    * number of the state tree.
    */
-  #metadata?: { version: number };
+  #metadata?: MetaData;
 
   #isExtensionInitialized: boolean = false;
 
-  #localStore: ExtensionStore | ReadOnlyNetworkStore;
+  #localStore: BaseStore;
+
+  #db: DB;
+
+  #open: boolean = false;
 
-  constructor({
-    localStore,
-  }: {
-    localStore: ExtensionStore | ReadOnlyNetworkStore;
-  }) {
+  constructor({ localStore }: { localStore: BaseStore }) {
     this.#localStore = localStore;
+    this.#db = new DB();
+  }
+
+  async open() {
+    if (!this.#open) {
+      await this.#db.open('metamask-vault', 1);
+      this.#open = true;
+    }
   }
 
-  setMetadata(metadata: { version: number }) {
+  setMetadata(metadata: MetaData) {
     this.#metadata = metadata;
   }
 
+  /**
+   * Sets the state in the local store.
+   *
+   * @param state - The state to set in the local store. This should be an object
+   * containing the state data to be stored.
+   * @throws Error if the state is missing or if the metadata is not set before
+   * calling this method.
+   * @throws Error if the local store is not open.
+   * @throws Error if the data persistence fails during the write operation.
+   */
   async set(state: MetaMaskStateType) {
+    await this.open();
+
     if (!state) {
       throw new Error('MetaMask - updated state is missing');
     }
-    if (!this.#metadata) {
+    const meta = this.#metadata;
+    if (!meta) {
       throw new Error('MetaMask - metadata must be set before calling "set"');
     }
-    try {
-      await this.#localStore.set({ data: state, meta: this.#metadata });
-      if (this.#dataPersistenceFailing) {
-        this.#dataPersistenceFailing = false;
-      }
-    } catch (err) {
-      if (!this.#dataPersistenceFailing) {
-        this.#dataPersistenceFailing = true;
-        captureException(err);
-      }
-      log.error('error setting state in local store:', err);
-    } finally {
-      this.#isExtensionInitialized = true;
-    }
+
+    await navigator.locks.request(
+      STATE_LOCK,
+      { mode: 'exclusive' },
+      async () => {
+        try {
+          // atomically set all the keys
+          await this.#localStore.set({
+            data: state,
+            meta,
+          });
+
+          // back up the vault in IndexedDB
+          const keyringController =
+            state.KeyringController as KeyringControllerState;
+          const vault = keyringController?.vault;
```

Comment:
Nit: We can avoid the cast by doing something like this (using `isObject` and `hasProperty` from `@metamask/utils`)

```suggestion
          const keyringController = state.KeyringController;
          const vault =
            isObject(keyringController) &&
            hasProperty(keyringController, 'vault') &&
            typeof keyringController.vault === 'string'
              ? keyringController.vault
              : null;
```

Quite verbose, but it's more correct in terms of the types. We don't know that the state here is actually shaped as the state type dictates. Even absent an error, we might have just added a new property that's missing from persisted state.